### PR TITLE
Add NVMe disk controller support for hardware version >=13 VM. (#43000)

### LINKED ISSUE #43000: Request for NVMe controller support
##### SUMMARY
From ESXi 6.5, NVMe controller is supported. We want to create new VM with disk controller set to NVMe.

##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
vmware_guest

##### ANSIBLE VERSION
```
ansible 2.6.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vmware/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]

```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
ESXi 6.5 and later.

##### STEPS TO REPRODUCE
Add support "vim.vm.device.VirtualNVMEController()" virtual device for hardware version >= 13.
```yaml
      hardware:
        version: 13
        memory_mb: 4096
        num_cpus: 2
        num_cpu_cores_per_socket: 1
        scsi: lsilogicsas
        nvme: true
      disk:
      - size_gb: 40
         type: thin
         controller: nvme
         datastore: datastore1
      - size_gb: 4
         type: thin
         controller: scsi
         datastore: datastore1
```

##### EXPECTED RESULTS
NVMe controller can be added in the hardware list, and disk in the list can specify attaching to which controller.

##### ACTUAL RESULTS
N/A

### PR DESCRIPTION
**Pull request by Tomorrow9**
##### SUMMARY
Fixes #43000 

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
vmware_guest

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (testing_PR43000 605c748d4a) last updated 2018/08/01 15:09:25 (GMT +800)
  config file = /root/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/ansible/lib/ansible
  executable location = /root/ansible/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
```

##### ADDITIONAL INFORMATION

Add a new "hardware" disk controller type besides "scsi", named "nvme".

```

```

### CODE REVIEW

**Comment by Akasurde** [comment]
@Tomorrow9 Thanks for the PR. Could you please write tests for this change ? Thanks.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -649,6 +665,28 @@ def create_scsi_disk(self, scsi_ctl, disk_index=None):
...
+    def create_nvme_disk(self, disk_ctl_key, disk_index=None):
+        diskspec = vim.vm.device.VirtualDeviceSpec()
+        diskspec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
+        diskspec.fileOperation = vim.vm.device.VirtualDeviceSpec.FileOperation.create
+        diskspec.device = vim.vm.device.VirtualDisk()
+        diskspec.device.backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
+        diskspec.device.controllerKey = disk_ctl_key
+        # one nvme controller attach 0 - 14 disks
+        if self.next_disk_unit_number > 14:
+            raise AssertionError("unitNumber for nvme disk >14, valid 0-14.")
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: Rather than raising error, please use `module.fail_json`

**Comment by Tomorrow9** [comment]
> @Tomorrow9 This PR was evaluated as a potentially problematic PR for the following reasons:
> 
> * More than 50 changed files.
> * More than 50 commits.
> 
> Such PR can only be merged by human. Contact a Core team member to review this PR on IRC: `#ansible-devel` on irc.freenode.net
Sorry, I made a mistake, revert to the last good state.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 2]
```diff
@@ -1669,8 +1737,9 @@ def configure_disks(self, vm_obj):
             # increment index for next disk search
             disk_index += 1
             # index 7 is reserved to SCSI controller
-            if disk_index == 7:
-                disk_index += 1
+            if 'nvme' not in self.params['hardware']:
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1740]: if `nvme` is in the hardware, the disk_index is allowed to be 7?

**Review by pdellaert — APPROVED** [review]
Overall LGTM
nice addition, also really good to add the integration tests :). 

One question/comment, but that's more for me to learn.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 2 above)
Comment (Tomorrow9) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py:1740]: Yeah, It can be 0 to 14.

**Comment by Tomorrow9** [comment]
Hi, this submit https://github.com/ansible/ansible/pull/49534 contains the "nvme" controller support, if it's merged I think changes in this PR may not required or need some modifications. So I'll wait for #49534 to be merged and then think about this one. Thanks.

**Comment by Akasurde** [comment]
@Tomorrow9 Do you want to merge this with "multiple disk and controller" support PR ? Let me know.

**Comment by Tomorrow9** [comment]
> @Tomorrow9 Do you want to merge this with "multiple disk and controller" support PR ? Let me know.

Hi @Akasurde, sorry for not seeing your question. I'm still testing this PR #49534 after some changes, and yes I'm going to add NVME support in that "multiple disk and controller" support. Thanks.

**Comment by Akasurde** [comment]
Closing in favour of https://github.com/ansible/ansible/pull/49534

**Pull request closed without merging**